eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -0,0 +1,255 @@
+  type Environment (ConwayDELEG era) = DelegEnv era
+  type BaseM (ConwayDELEG era) = ShelleyBase
+  type PredicateFailure (ConwayDELEG era) = ConwayDelegPredFailure era
+  type Event (ConwayDELEG era) = ConwayDelegEvent era
+
+  transitionRules = [conwayDelegTransition @era]
+
+conwayDelegTransition :: 
+  forall era.
+  ( EraTx era
+  -- , ConwayEraTxBody era
+  -- , ShelleyEraTxBody era
+  -- , State (EraRule "DELEG" era) ~ CertState era
+  -- , Signal (EraRule "DELEG" era) ~ TxCert era
+  , TxCert era ~ ConwayTxCert era
+  -- , Environment (EraRule "DELEG" era) ~ DelegEnv era
+  -- , EraRule "DELEG" era ~ ConwayDELEG era
+  ) => 
+  TransitionRule (ConwayDELEG era)
+conwayDelegTransition = do
+  TRC 
+    ( DelegEnv _slot ptr _acnt pp
+    , cState@CertState
+        { certVState = VState{vsDReps}
+        , certPState = _
+        , certDState = certDState@DState{dsUnified} 
+        } 
+    , c
+    ) <- judgmentContext
+  let pd = pp ^. ppKeyDepositL
+  case c of
+    ConwayTxCertDeleg (ConwayRegCert stakeCred sMayCoin) -> do
+      sMayCoin == SNothing || sMayCoin == SJust pd
+        ?! RegCertProvidedCoinNotEqualToExpectedPoolDepositDELEG sMayCoin
+      UM.notMember stakeCred (RewDepUView dsUnified) 
+        ?! RegCertStakeKeyAlreadyRegisteredDELEG stakeCred
+      let ppDeposit = UM.compactCoinOrError pd
+          newRDs = UM.RewDepUView dsUnified UM.∪ (stakeCred, UM.RDPair (UM.CompactCoin 0) ppDeposit)
+          newUMap = UM.PtrUView newRDs UM.∪ (ptr, stakeCred)
+      pure $ 
+        cState 
+          { certDState = 
+              certDState 
+                { dsUnified = newUMap }
+          }
+    ConwayTxCertDeleg (ConwayUnRegCert stakeCred sMayCoin) -> do
+      sMayCoin == SNothing || sMayCoin == SJust pd
+        ?! UnRegCertProvidedCoinNotEqualToExpectedPoolDepositDELEG sMayCoin
+      UM.member stakeCred (RewDepUView dsUnified) 
+        ?! UnRegCertStakeKeyNotRegisteredDELEG stakeCred
+      let mRDPair = UM.lookup stakeCred (RewDepUView dsUnified)
+          mRewardCoin = UM.rdReward <$> mRDPair
+          mDepositCoin = UM.rdDeposit <$> mRDPair
+      mRewardCoin == Just mempty 
+        ?! UnRegCertStakeKeyNonZeroAccountBalanceDELEG (UM.fromCompact <$> mRewardCoin)
+      maybeToStrictMaybe (UM.fromCompact <$> mDepositCoin) == sMayCoin
+        ?! UnRegCertStakeKeyDepositDoesNotMatchExpectedDepositDELEG (UM.fromCompact <$> mDepositCoin)
+      let umRDRemoved = Set.singleton stakeCred UM.⋪ UM.RewDepUView dsUnified
+          umSPoolRemoved = Set.singleton stakeCred UM.⋪ UM.SPoolUView umRDRemoved
+          newUMap = UM.PtrUView umSPoolRemoved UM.⋫ Set.singleton stakeCred
+      pure $
+        cState
+          { certDState =
+              certDState
+                { dsUnified = newUMap }
+          }
+    ConwayTxCertDeleg (ConwayDelegCert stakeCred delegatee) -> do
+      UM.member stakeCred (RewDepUView dsUnified)
+        ?! DelegCertStakeKeyNotRegisteredDELEG stakeCred
+      case delegatee of
```

Comment:
I would extract this as a function and then make it so that the `DelegStakeVote` case just calls the same function with `DelegStake` and `DelegVote` and then combines the results together. That way we can get rid of some code duplication.

Reply:
Great suggestion!